rmw_dds_common/src/qos.cpp

```diff
@@ -0,0 +1,313 @@
+    // Reliability for publisher is "system default" and subscription is reliable
+    if (publisher_qos.reliability == RMW_QOS_POLICY_RELIABILITY_SYSTEM_DEFAULT &&
+      subscription_qos.reliability == RMW_QOS_POLICY_RELIABILITY_RELIABLE)
+    {
+      *compatibility = RMW_QOS_COMPATIBILITY_WARNING;
+      rmw_ret_t append_ret = _append_to_buffer(
+        "WARNING: Reliable subscription, but publisher is system default;",
+        reason,
+        reason_size);
+      if (RMW_RET_OK != append_ret) {
+        return append_ret;
+      }
+    } else {
+      // Reliability for publisher is best effort and subscription is "system default"
+      if (publisher_qos.reliability == RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT &&
+        subscription_qos.reliability == RMW_QOS_POLICY_RELIABILITY_SYSTEM_DEFAULT)
+      {
+        *compatibility = RMW_QOS_COMPATIBILITY_WARNING;
+        rmw_ret_t append_ret = _append_to_buffer(
+          "WARNING: Best effort publisher, but subscription is system default;",
+          reason,
+          reason_size);
+        if (RMW_RET_OK != append_ret) {
+          return append_ret;
+        }
+      }
+    }
+
+    // Durability for publisher is "system default" and subscription is transient local
+    if (publisher_qos.durability == RMW_QOS_POLICY_DURABILITY_SYSTEM_DEFAULT &&
+      subscription_qos.durability == RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL)
+    {
+      *compatibility = RMW_QOS_COMPATIBILITY_WARNING;
+      rmw_ret_t ret = _append_to_buffer(
+        "WARNING: Transient local subscription, but publisher is system default;",
+        reason,
+        reason_size);
+      if (RMW_RET_OK != ret) {
+        return ret;
+      }
+    } else {
+      // Durability for publisher is volatile and subscription is "system default"
+      if (publisher_qos.durability == RMW_QOS_POLICY_DURABILITY_VOLATILE &&
+        subscription_qos.durability == RMW_QOS_POLICY_DURABILITY_SYSTEM_DEFAULT)
+      {
+        *compatibility = RMW_QOS_COMPATIBILITY_WARNING;
+        rmw_ret_t ret = _append_to_buffer(
+          "WARNING: Volatile publisher, but subscription is system default;",
+          reason,
+          reason_size);
+        if (RMW_RET_OK != ret) {
+          return ret;
+        }
+      }
+    }
+
+    // Automatic liveliness for publisher and "system default" for subscription
+    if (publisher_qos.liveliness == RMW_QOS_POLICY_LIVELINESS_AUTOMATIC &&
+      subscription_qos.liveliness == RMW_QOS_POLICY_LIVELINESS_SYSTEM_DEFAULT)
+    {
+      *compatibility = RMW_QOS_COMPATIBILITY_WARNING;
+      rmw_ret_t ret = _append_to_buffer(
+        "WARNING: Publisher's liveliness is automatic, but subscription's is system default;",
+        reason,
+        reason_size);
+      if (RMW_RET_OK != ret) {
+        return ret;
+      }
+    } else {
+      if (publisher_qos.liveliness == RMW_QOS_POLICY_LIVELINESS_SYSTEM_DEFAULT &&
+        subscription_qos.liveliness == RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_TOPIC)
```

Comment:
I just realized there's another case where we don't know if the profiles are valid, if both the publisher and subscription are "system default".

Reply:
Maybe I'm overthinking this one. If both are system default, then presumably they are using the same underlying QoS settings, though I guess it's possible the policy for a subscription can still be configured independently from the publisher s.t. they are incompatible (that would be unfortunate).

Reply:
> then presumably they are using the same underlying QoS settings

The system default is read from a qos profile file (if there is one), if not it's the DDS defaults.
You could potentially write a qos profile file where the default profile for publishers is incompatible with the one for subscriptions, but that would be bizarre ...

Reply:
> You could potentially write a qos profile file where the default profile for publishers is incompatible with the one for subscriptions, but that would be bizarre ...

Yeah, I agree; bizarre. But, given this information we can't do any better than say there is a possible compatibility issue, so I think the best thing to do is warn in this case.

Reply:
Handled the case where we have two "system default" / "unknown" values: 6a67320